docs/concepts/architecture/cloud-controller.md

```diff
@@ -0,0 +1,249 @@
+Cloud Controller Manager
+-------------------------
+
+The Cloud Controller Manager (CCM) was originally created to allow cloud specific vendor code and kubernetes core to evolve independent of one another. The cloud controller manager runs alongside other master components such as Kube Controller Manager, Kube Apiserver,and Kube Scheduler.
+
+Cloud Controller Manager's design is based on a plugin mechanism that allows new cloud providers to integrate with Kubernetes easily using plugins. There are distinct plans in place for on-boarding new cloud providers on Kubernetes, and for migrating new cloud provider from the old model to using the CCM model.
+
+In this document, the concepts behind the components of cloud controller manager, and its associated functions will be discussed in detail. Here's the architecture of a Kubernetes cluster without the Cloud Controller Manager
+
+![Pre CCM Kube Arch](https://i.imgur.com/LAmMku4.png)
+
+Design
+-------
+
+In the architecture diagram of a Kubernetes cluster above, Kubernetes and the Cloud integrate from many different components.Namely
+
+1. Kubelet
+2. Kube Controller Manager
+3. Kube ApiServer
+
+The CCM consolidates all of these cloud dependent logic from the above 3 components to create a single point of integration with the cloud. The new architecture with CCM looks like this:
+
+![CCM Kube Arch](https://i.imgur.com/iqiIFHY.png)
+
+Components of CCM
+------------------
+The CCM breaks away some of the functionality of Kube Controller Manager (KCM) and runs it as a separate process. Specifically, it breaks away those controllers in KCM that are cloud dependent. The KCM has the following cloud dependent controller loops:
+
+ 1. Node Controller
+ 2. Volume Controller
+ 3. Route Controller
+ 4. Service Controller
+
+The CCM currently (v1.8) runs the following controllers from the above list
+
+1. Node Controller
+2. Route Controller 
+3. Service Controller
+
+Additionally, it runs another controller called the PersistentVolumeLabels controller. This controller is responsible for setting the zone and region labels on PersistentVolumes created in GCE and AWS clouds. 
+
+To summarize,the CCM runs the following 4 controllers:
+
+1. Node Controller
+2. Route Controller 
+3. Service Controller
+4. Persistent Volume Labels Controller
+
+**Note:** Volume Controller was deliberately chosen to *NOT* be a part of CCM. Due to the complexity involved and due to the existing efforts to abstract away vendor specific volume logic, it was decided that Volume Controller will not be moved to CCM.   
```

Comment:
I'm a bit confused here. The KCM list doesn't include "PV Labels Controller", but it is listed as one that CCM runs. Is the "PV Labels Controller" a new control loop that has been refactored out from the PV Controller.

If not, does it makes sense to go ahead and just list "PV Labels Controler" under both the KCM and CCM lists for further clarity?

Reply:
>it was decided that Volume Controller will not be moved to CCM.

What does this mean? Does this mean that volumes don't work at all in my cluster? Does it mean I should still be enabling the cloud provider on the `kubelet`s so they can fulfill this role?

Reply:
@colemickens This is why I had the original comment on this PR to add a doc about our transition plan. 

The Volume functionality will not be lost. @cheftako is working on a PR to enable volume controller to run despite having the CCM. https://github.com/kubernetes/kubernetes/pull/52371

In case of PVL controller, it was added into the CCM, it did not exist before CCM was created. This was done to move the PV labelling logic in kube-apiserver (it was an admission controller) to the CCM. It does not run on the KCM. It is a CCM only controller.

Reply:
Please say this more clearly here in the doc

Reply:
done